pkg/kubectl/cmd/set/set_volume.go

```diff
@@ -0,0 +1,958 @@
+	cmd.Flags().StringVar(&addOpts.ClaimMode, "claim-mode", "ReadWriteOnce", "Set the access mode of the claim to be created. Valid values are ReadWriteOnce (rwo), ReadWriteMany (rwm), or ReadOnlyMany (rom)")
+	cmd.Flags().StringVar(&addOpts.Source, "source", "", "Details of volume source as json string. This can be used if the required volume type is not supported by --type option. (e.g.: '{\"gitRepo\": {\"repository\": <git-url>, \"revision\": <commit-hash>}}')")
+
+	cmdutil.AddDryRunFlag(cmd)
+	cmdutil.AddPrinterFlags(cmd)
+	cmd.MarkFlagFilename("filename", "yaml", "yml", "json")
+
+	// deprecate --list option
+	cmd.Flags().MarkDeprecated("list", "Volumes and volume mounts can be listed by providing a resource with no additional options.")
+
+	return cmd
+}
+
+func (o *VolumeOptions) Validate(cmd *cobra.Command, args []string) error {
+	if len(o.Selector) > 0 {
+		if _, err := labels.Parse(o.Selector); err != nil {
+			return errors.New("--selector=<selector> must be a valid label selector")
+		}
+		if o.All {
+			return errors.New("you may specify either --selector or --all but not both")
+		}
+	}
+	if len(o.FilenameOptions.Filenames) == 0 && len(args) < 1 {
+		return errors.New("provide one or more resources to add, list, or delete volumes on as TYPE/NAME")
+	}
+
+	numOps := 0
+	if o.Add {
+		numOps++
+	}
+	if o.Remove {
+		numOps++
+	}
+	if o.List {
+		numOps++
+	}
+
+	switch {
+	case numOps == 0:
+		o.List = true
+	case numOps > 1:
+		return errors.New("you may only specify one operation at a time")
+	}
+
+	output := cmdutil.GetFlagString(cmd, "output")
+	if o.List && len(output) > 0 {
+		return errors.New("--list and --output may not be specified together")
+	}
+
+	err := o.AddOpts.Validate(o.Add)
+	if err != nil {
+		return err
+	}
+	// Removing all volumes for the resource type needs confirmation
+	if o.Remove && len(o.Name) == 0 && !o.Confirm {
+		return errors.New("must provide --confirm for removing more than one volume")
+	}
+	return nil
+}
+
+func (a *AddVolumeOptions) Validate(isAddOp bool) error {
+	if isAddOp {
+		if len(a.Type) == 0 && (len(a.ClaimName) > 0 || len(a.ClaimSize) > 0) {
+			a.Type = "persistentvolumeclaim"
+			a.TypeChanged = true
+		}
+		if len(a.Type) == 0 && (len(a.SecretName) > 0) {
+			a.Type = "secret"
+			a.TypeChanged = true
+		}
+		if len(a.Type) == 0 && (len(a.ConfigMapName) > 0) {
+			a.Type = "configmap"
+			a.TypeChanged = true
+		}
+		if len(a.Type) == 0 && (len(a.Path) > 0) {
+			a.Type = "hostpath"
+			a.TypeChanged = true
+		}
+		if len(a.Type) == 0 {
+			a.Type = "emptydir"
+		}
+
+		if len(a.Type) == 0 && len(a.Source) == 0 {
+			return errors.New("must provide --type or --source for --add operation")
+		} else if a.TypeChanged && len(a.Source) > 0 {
+			return errors.New("either specify --type or --source but not both for --add operation")
+		}
+
+		if len(a.Type) > 0 {
```

Comment:
I wonder shall we split these checking to separate methods to make it more concise and maintainable.

Reply:
@dixudx OK